src/iso15118/d20/state/dc_charge_loop.cpp

```diff
@@ -192,28 +191,43 @@ handle_request(const message_20::DC_ChargeLoopRequest& req, const d20::Session&
             return {response_with_code(res, message_20::ResponseCode::FAILED), ev_feedback_control_mode};
         }
 
-        auto& mode = res.control_mode.emplace<Scheduled_BPT_DC_Res>();
-        convert(mode, dc_limits);
-
         const auto& req_mode = std::get<Scheduled_BPT_DC_Req>(req.control_mode);
-        auto& scheduled_values = ev_feedback_control_mode.emplace<session::feedback::DcChargeScheduledMode>();
-        convert(scheduled_values, req_mode);
+        auto& req_scheduled_values = ev_feedback_control_mode.emplace<session::feedback::DcChargeScheduledMode>();
+        convert(req_scheduled_values, req_mode);
+
+        auto& res_mode = res.control_mode.emplace<Scheduled_BPT_DC_Res>();
+        convert(res_mode, dc_limits);
 
     } else if (std::holds_alternative<Dynamic_DC_Req>(req.control_mode)) {
-        if (session.get_selected_energy_service() != message_20::ServiceCategory::DC) {
+
+        if (session.get_selected_control_mode() != message_20::ControlMode::Dynamic or
+            session.get_selected_energy_service() != message_20::ServiceCategory::DC) {
             return {response_with_code(res, message_20::ResponseCode::FAILED), ev_feedback_control_mode};
         }
 
-        auto& mode = res.control_mode.emplace<Dynamic_DC_Res>();
-        convert(mode, dc_limits);
-
         const auto& req_mode = std::get<Dynamic_DC_Req>(req.control_mode);
-        auto& dynamic_values = ev_feedback_control_mode.emplace<session::feedback::DcChargeDynamicMode>();
-        convert(dynamic_values, req_mode);
+        auto& req_dynamic_values = ev_feedback_control_mode.emplace<session::feedback::DcChargeDynamicMode>();
+        convert(req_dynamic_values, req_mode);
+
+        auto& res_mode = res.control_mode.emplace<Dynamic_DC_Res>();
+        convert(res_mode, dc_limits);
+
+        if (session.get_selected_mobility_needs_mode() == message_20::MobilityNeedsMode::ProvidedBySecc) {
+            if (new_departure_time.has_value()) {
```

Comment:
This is also personal taste, but with optionals I prefer to just use `if (new_depature_time)`

Reply:
Also note here, 2 indents could be saved by having a helper function so that you can write
```
res_mode.departure_time = helper_function(new_departure_time, res.header.timestamp)
```

Reply:
I added a helper template function for this